[MediaStream] Add support for ImageCapture

eric-carlson opened the pull request:
#### 3f6e79379d24178b2b960edee30ea00d4ec792f2
<pre>
[MediaStream] Add support for ImageCapture
<a href="https://bugs.webkit.org/show_bug.cgi?id=262412">https://bugs.webkit.org/show_bug.cgi?id=262412</a>
rdar://116261148

Reviewed by Jean-Yves Avenard.

Add the ImageCapture object, implement the getPhotoCapabilities method.

* LayoutTests/fast/mediastream/image-capture-creation-expected.txt: Added.
* LayoutTests/fast/mediastream/image-capture-creation.html: Added.
* LayoutTests/fast/mediastream/image-capture-get-photo-capabilities-expected.txt: Added.
* LayoutTests/fast/mediastream/image-capture-get-photo-capabilities.html: Added.
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/mediastream/FillLightMode.idl: Copied from Source/WebCore/Modules/mediastream/MeteringMode.h.
* Source/WebCore/Modules/mediastream/ImageCapture.cpp: Added.
(WebCore::ImageCapture::create):
(WebCore::ImageCapture::ImageCapture):
(WebCore::ImageCapture::getPhotoCapabilities):
(WebCore::ImageCapture::activeDOMObjectName const):
* Source/WebCore/Modules/mediastream/ImageCapture.h: Added.
(WebCore::ImageCapture::track const):
* Source/WebCore/Modules/mediastream/ImageCapture.idl: Added.
* Source/WebCore/Modules/mediastream/MediaSettingsRange.idl: Copied from Source/WebCore/Modules/mediastream/MeteringMode.h.
* Source/WebCore/Modules/mediastream/MediaStreamTrack.cpp:
(WebCore::MediaStreamTrack::getPhotoCapabilities const):
* Source/WebCore/Modules/mediastream/MediaStreamTrack.h:
* Source/WebCore/Modules/mediastream/PhotoCapabilities.idl: Copied from Source/WebCore/Modules/mediastream/MeteringMode.h.
* Source/WebCore/Modules/mediastream/RedEyeReduction.idl: Copied from Source/WebCore/Modules/mediastream/MeteringMode.h.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/platform/mediastream/FillLightMode.h:
* Source/WebCore/platform/mediastream/MediaSettingsRange.h:
* Source/WebCore/platform/mediastream/MediaStreamTrackPrivate.cpp:
(WebCore::MediaStreamTrackPrivate::getPhotoCapabilities):
* Source/WebCore/platform/mediastream/MediaStreamTrackPrivate.h:
* Source/WebCore/platform/mediastream/MeteringMode.h:
* Source/WebCore/platform/mediastream/PhotoCapabilities.h:
* Source/WebCore/platform/mediastream/RedEyeReduction.h:
* Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.h:
* Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.mm:
(WebCore::AVVideoCaptureSource::getPhotoCapabilities):
* Source/WebCore/platform/mock/MockRealtimeVideoSource.cpp:
(WebCore::MockRealtimeVideoSource::getPhotoCapabilities):
* Source/WebCore/platform/mock/MockRealtimeVideoSource.h:
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.cpp:
(WebKit::UserMediaCaptureManagerProxy::SourceProxy::getPhotoCapabilities):
(WebKit::UserMediaCaptureManagerProxy::getPhotoCapabilities):
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.messages.in:
* Source/WebKit/WebProcess/cocoa/RemoteRealtimeMediaSource.cpp:
(WebKit::RemoteRealtimeMediaSource::getPhotoCapabilities):
* Source/WebKit/WebProcess/cocoa/RemoteRealtimeMediaSource.h:
* Source/WebKit/WebProcess/cocoa/RemoteRealtimeMediaSourceProxy.cpp:
(WebKit::RemoteRealtimeMediaSourceProxy::getPhotoCapabilities):
* Source/WebKit/WebProcess/cocoa/RemoteRealtimeMediaSourceProxy.h:

Canonical link: <a href="https://commits.webkit.org/268732@main">https://commits.webkit.org/268732@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f641c8cb8a48ba6ece11a5cb823f4ffbdc01b066

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/20456 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/20883 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/21527 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/22355 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/19087 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/20688 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/24141 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/21057 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/20487 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/20676 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/20545 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/17784 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/23207 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/20643 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/17713 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/18589 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/24875 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/18790 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/18765 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/22815 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/19356 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/16439 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/18569 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/18425 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4923 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/22909 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/19177 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on Source/WebCore/Modules/mediastream/ImageCapture.idl:
```diff
@@ -0,0 +1,43 @@
+
+    Promise<PhotoCapabilities> getPhotoCapabilities();
+
+    // Promise<Blob> takePhoto(optional PhotoSettings photoSettings = {});
```
TODO with linked to bug?

eric-carlson replied:
Will do

jyavenard commented on Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.cpp:
```diff
@@ -521,6 +526,12 @@ void UserMediaCaptureManagerProxy::clone(RealtimeMediaSourceIdentifier clonedID,
+void UserMediaCaptureManagerProxy::getPhotoCapabilities(RealtimeMediaSourceIdentifier sourceID, GetPhotoCapabilitiesCallback&& handler)
+{
+    if (auto* proxy = m_proxies.get(sourceID))
+        proxy->getPhotoCapabilities(WTFMove(handler));
```
need to call the handler with error here in case there's no proxy

eric-carlson replied:
Oh, good point!

jyavenard reviewed (approved):
LGTM

eric-carlson committed:
[MediaStream] Add support for ImageCapture
https://bugs.webkit.org/show_bug.cgi?id=262412
rdar://116261148

Reviewed by Jean-Yves Avenard.

Add the ImageCapture object, implement the getPhotoCapabilities method.

* LayoutTests/fast/mediastream/image-capture-creation-expected.txt: Added.
* LayoutTests/fast/mediastream/image-capture-creation.html: Added.
* LayoutTests/fast/mediastream/image-capture-get-photo-capabilities-expected.txt: Added.
* LayoutTests/fast/mediastream/image-capture-get-photo-capabilities.html: Added.
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/mediastream/FillLightMode.idl: Copied from Source/WebCore/Modules/mediastream/MeteringMode.h.
* Source/WebCore/Modules/mediastream/ImageCapture.cpp: Added.
(WebCore::ImageCapture::create):
(WebCore::ImageCapture::ImageCapture):
(WebCore::ImageCapture::getPhotoCapabilities):
(WebCore::ImageCapture::activeDOMObjectName const):
* Source/WebCore/Modules/mediastream/ImageCapture.h: Added.
(WebCore::ImageCapture::track const):
* Source/WebCore/Modules/mediastream/ImageCapture.idl: Added.
* Source/WebCore/Modules/mediastream/MediaSettingsRange.idl: Copied from Source/WebCore/Modules/mediastream/MeteringMode.h.
* Source/WebCore/Modules/mediastream/MediaStreamTrack.cpp:
(WebCore::MediaStreamTrack::getPhotoCapabilities const):
* Source/WebCore/Modules/mediastream/MediaStreamTrack.h:
* Source/WebCore/Modules/mediastream/PhotoCapabilities.idl: Copied from Source/WebCore/Modules/mediastream/MeteringMode.h.
* Source/WebCore/Modules/mediastream/RedEyeReduction.idl: Copied from Source/WebCore/Modules/mediastream/MeteringMode.h.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/platform/mediastream/FillLightMode.h:
* Source/WebCore/platform/mediastream/MediaSettingsRange.h:
* Source/WebCore/platform/mediastream/MediaStreamTrackPrivate.cpp:
(WebCore::MediaStreamTrackPrivate::getPhotoCapabilities):
* Source/WebCore/platform/mediastream/MediaStreamTrackPrivate.h:
* Source/WebCore/platform/mediastream/MeteringMode.h:
* Source/WebCore/platform/mediastream/PhotoCapabilities.h:
* Source/WebCore/platform/mediastream/RedEyeReduction.h:
* Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.h:
* Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.mm:
(WebCore::AVVideoCaptureSource::getPhotoCapabilities):
* Source/WebCore/platform/mock/MockRealtimeVideoSource.cpp:
(WebCore::MockRealtimeVideoSource::getPhotoCapabilities):
* Source/WebCore/platform/mock/MockRealtimeVideoSource.h:
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.cpp:
(WebKit::UserMediaCaptureManagerProxy::SourceProxy::getPhotoCapabilities):
(WebKit::UserMediaCaptureManagerProxy::getPhotoCapabilities):
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.messages.in:
* Source/WebKit/WebProcess/cocoa/RemoteRealtimeMediaSource.cpp:
(WebKit::RemoteRealtimeMediaSource::getPhotoCapabilities):
* Source/WebKit/WebProcess/cocoa/RemoteRealtimeMediaSource.h:
* Source/WebKit/WebProcess/cocoa/RemoteRealtimeMediaSourceProxy.cpp:
(WebKit::RemoteRealtimeMediaSourceProxy::getPhotoCapabilities):
* Source/WebKit/WebProcess/cocoa/RemoteRealtimeMediaSourceProxy.h:

Canonical link: https://commits.webkit.org/268732@main
